ext/nnstreamer/android_source/gstamcsrc.c

```diff
@@ -0,0 +1,971 @@
+      "width", G_TYPE_INT, priv->width,
+      "height", G_TYPE_INT, priv->height,
+      "interlaced", G_TYPE_BOOLEAN, FALSE,
+      "pixel-aspect-ratio", GST_TYPE_FRACTION, 1, 1,
+      "framerate", GST_TYPE_FRACTION, priv->framerate, 1, NULL);
+
+  gst_caps_append_structure (caps, format);
+
+  gst_structure_free (format);
+
+  GST_OBJECT_UNLOCK (self);
+
+  return caps;
+}
+
+/**
+ * @brief set a buffer which is a head item in data queue
+ */
+static GstFlowReturn
+gst_amc_src_create (GstPushSrc * src, GstBuffer ** buffer)
+{
+  GstAMCSrc *self = GST_AMC_SRC_CAST (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+  GstDataQueueItem *item;
+
+  if (!gst_data_queue_pop (priv->outbound_queue, &item)) {
+    GST_DEBUG_OBJECT (self, "We're flushing");
+    return GST_FLOW_FLUSHING;
+  }
+
+  *buffer = GST_BUFFER (item->object);
+  g_free (item);
+
+  return GST_FLOW_OK;
+}
+
+/**
+ * @brief change state function for this element
+ * each trainsition sends the corresponding message to a looper
+ */
+static GstStateChangeReturn gst_amc_src_change_state (GstElement * element, 
+    GstStateChange transition)
+{
+  GstAMCSrc *self;
+  GstAMCSrcPrivate *priv;
+  GstStateChangeReturn ret;
+  gboolean buffer_state_change_success = TRUE;
+  gchar *dirname = NULL;
+
+  self = GST_AMC_SRC (element);
+  priv  = GST_AMC_SRC_GET_PRIVATE (self); 
+
+  switch (transition) {
+    case GST_STATE_CHANGE_PAUSED_TO_PLAYING:
+      Looper_post (priv->looper, MSG_CODEC_RESUME, self, false);
+      LOGI ("PAUSED => PLAYING");
+      break;
+    default:
+      break;
+  }
+
+  ret = GST_ELEMENT_CLASS (parent_class)->change_state (element, transition);
+
+  switch (transition) {
+    case GST_STATE_CHANGE_PLAYING_TO_PAUSED:
+      Looper_post (priv->looper, MSG_CODEC_PAUSE, self, false);
+      LOGI ("PLAYING => PAUSED");
+      break;
+    default:
+      break;
+  }
+
+  if (buffer_state_change_success == FALSE) 
+    ret = GST_STATE_CHANGE_FAILURE;
+
+  return ret;
+}
+
+/**
+ * @brief destroy callback for a data queue item
+ */
+static void
+data_queue_item_free (GstDataQueueItem * item)
+{
+  g_clear_pointer (&item->object, (GDestroyNotify) gst_mini_object_unref);
+  g_free (item);
+}
+
+/**
+ * @brief feed a decoded data from media codec to pipeline
+ * @note it avoid memcpy() by wrapping memory
+ * @param[in] buf codec's output buffer
+ * @param[in] idx codec's output buffer idx
+ * @param[in] real_size codec's output data size
+ * @param[in] buf_size codec's output buffer size
+ */
+static void
+feed_frame_buf (GstAMCSrc *self, guint8 *buf, gint idx, gsize real_size, gsize buf_size)
+{
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  GstBuffer *buffer;
+  GstMemory *mem;
+  GstDataQueueItem *item;
+  GstClockTime duration = GST_CLOCK_TIME_NONE;
+  GstClockTime current_ts = GST_CLOCK_TIME_NONE;
+  GstClock *clock;
+
+  GstWrappedBuf *wrapped_buf;
+
+  if ((clock = GST_ELEMENT_CLOCK (self))) {
```

Comment:
consider to get referenced ```clock = gst_element_get_clock (GST_ELEMENT (self))``` and remove gst_object_ref (clock);

Reply:
OK. I'll check this.